pkg/clustermesh/mcsapi/mcsapiservicespec.go

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package mcsapi
+
+import (
+	"encoding/json"
+	"path"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	mcsapiv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"github.com/cilium/cilium/pkg/kvstore"
+)
+
+var (
+	// ServiceExportStorePrefix is the kvstore prefix of the shared store
+	//
+	// WARNING - STABLE API: Changing the structure or values of this will
+	// break backwards compatibility
+	ServiceExportStorePrefix = path.Join(kvstore.BaseKeyPrefix, "state", "serviceexports", "v1")
+)
+
+type MCSAPIServiceSpec struct {
+	// Cluster is the cluster name the service is configured in
+	Cluster string `json:"cluster"`
+
+	// Name is the name of the ServiceImport or ServiceExport parent resources.
+	Name string `json:"name"`
+
+	// Namespace is the cluster namespace the service is configured in
+	Namespace string `json:"namespace"`
+
+	// ExportCreationTimestamp is the timestamp representing when the
+	// ServiceExport object was created. It is used for conflict resolution.
+	ExportCreationTimestamp metav1.Time `json:"exportCreationTimestamp"`
+
+	// Ports are the list of ports of the Service in MCS API format
+	Ports []mcsapiv1alpha1.ServicePort `json:"ports"`
+
+	// Type defines the type of this service.
+	// Must be ClusterSetIP or Headless.
+	Type mcsapiv1alpha1.ServiceImportType `json:"type"`
+
+	// Supports "ClientIP" and "None". Used to maintain session affinity.
+	// Enable client IP based session affinity.
+	// Must be ClientIP or None.
+	// Defaults to None.
+	// Ignored when type is Headless
+	// More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
+	SessionAffinity corev1.ServiceAffinity `json:"sessionAffinity,omitempty"`
+
+	// SessionAffinityConfig contains session affinity configuration.
+	SessionAffinityConfig *corev1.SessionAffinityConfig `json:"sessionAffinityConfig,omitempty"`
+}
+
+// GetKeyName returns the kvstore key to be used for MCSAPIServiceSpec
+func (s *MCSAPIServiceSpec) GetKeyName() string {
+	// WARNING - STABLE API: Changing the structure of the key may break
+	// backwards compatibility
+	return path.Join(s.Cluster, s.Namespace, s.Name)
+}
+
+// Marshal returns the MCS-API Service Spec object as JSON byte slice
+func (s *MCSAPIServiceSpec) Marshal() ([]byte, error) {
+	return json.Marshal(s)
+}
+
+// Unmarshal parses the JSON byte slice and updates the MCS-API Service Spec receiver
+func (s *MCSAPIServiceSpec) Unmarshal(_ string, data []byte) error {
```

Comment:
The unmarshal logic should perform basic validation to prevent processing incorrect data that may then cause a panic or unexpected behavior down the line. In this case, I'd suggest to validate that `Cluster`, `Namespace` and `Name` are not empty, that `Port` names are unique and `Port`  types are valid, and that the `Type` and `SessionAffinity` fields contain valid values. You can find an example for ClusterServices [here](https://github.com/cilium/cilium/blob/b30a3a941c54d5f74d0d1100936a74d340553476/pkg/service/store/store.go#L133-L164).

Additionally, when implementing the reader side, there should be additional validation to check that the cluster name matches the expected one, and that the kvstore key matches the namespace and name. Again, you can find an example for ClusterServices [here](https://github.com/cilium/cilium/blob/b30a3a941c54d5f74d0d1100936a74d340553476/pkg/service/store/store.go#L179-L223).